### PR TITLE
Experiment saving and resumption

### LINKED ISSUE #13: Ability to continue experiment/study progress
### Feature request
If a study or an experiment run is stopped prior to completing training, it would be nice to be able to continue the training process after picking up where it left off.

### PR DESCRIPTION
**Pull request by mshalm**
In #13, @ebianchi noted the value of restarting the often-lengthy training process in an experiment.

This PR implements saving and resuming experiments via the `dair_pll.experiment.TrainingState` dataclass.
Every epoch, `SupervisedLearningExperiment` saves this object to disk at `checkpoint.pt`, which includes

- Train/valid/test split as a list of indices, so training data can be re-apportioned correctly.
- Current learned system state; optimizer state; and epoch number so that the training process can be resumed at the right point.
- Best validation loss, as well as the learned system state that achieved it and the epoch it was achieved on, so that the training process can be properly stopped.

As part of correctly saving data, the functionality for generating simulation data has been separated from the process of loading from disk. This reflects the recent design change to consider multiple experiments as being based around the same dataset. Data generation also now saves both the raw trajectory from the underlying `System` (in `data/ground_truth`) _as well as_ the one with added measurement noise (in `data/learning`). Data generation functionality is now moved to `dataset_generation.py`.

In order to keep the ballooning size of `experiment.py` down, configuration files have been migrated to `experiment_config.py` . For reproducibility, full configurations are also now pickled in `config.pkl` .

File loading and saving has been cleaned up as well, and the current folder structure is:
```
- storage/
  |
  - {storage_name}/
    |
    - data/
      | 
      - ground_truth/
      - learning/
    - runs/
      |
      - {run_name}/
        |
        - urdfs/
        - trajectory.gif
        - statistics.pkl
        - config.pkl
        - checkpoint.pt
        - wandb/
      - {other_run_name}/
        |
        - urdfs/
        - trajectory.gif
        - statistics.pkl
        - config.pkl
        - checkpoint.pt
        - wandb/
    - studies/
      |
      - {study_name}/
        |
        - optimal_hyperparameters.json
```

### CODE REVIEW

File: dair_pll/dataset_generation.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        self.config = config
+
+    def generate(self) -> None:
+        """Simulate trajectories and write them to disk."""
+        config = self.config
+        n_pop = config.n_pop
+        ground_truth_dir = file_utils.ground_truth_data_dir(config.storage)
+        learning_dir = file_utils.learning_data_dir(config.storage)
+        n_on_disk = file_utils.get_trajectory_count(ground_truth_dir)
+        n_to_add = 30
```
Comment (ebianchi) [inline dair_pll/dataset_generation.py]: Is this just a convenient "batch size" for generating `n_to_add` trajectories at once?  Does this work better/faster than `n_to_add = max(n_pop - n_on_disk, 0)`?

File: dair_pll/dataset_management.py [hunk 2]
```diff
@@ -54,234 +42,224 @@ class DataConfig:
...
-        for i in range(first, last):
-            self.in_slices.append(traj[(i + 1 - in_window):(i + 1), :])
-            self.out_slices.append(traj[(i + 1):(i + 1 + out_window), :])
+        trajectory_duration = trajectory.shape[0]
+        first_time_index = self.config.t_skip
+        last_time_index = trajectory_duration - self.config.t_prediction
+        previous_states_length = self.config.t_history
+        future_states_length = self.config.t_prediction
+        assert first_time_index <= last_time_index
+        for time in range(first_time_index, last_time_index):
```
Comment (ebianchi) [inline dair_pll/dataset_management.py]: A small nit:  I always interpret the words "duration" and "time" to be in units of seconds(/days/years/etc).  Contrary to those expectations, here `trajectory_duration` is a number of time steps and `time` is an index number.  I'd personally prefer `trajectory_length` or `trajectory_time_steps`, and `time_index` or `index` just so the units are extra clear.

File: dair_pll/dataset_management.py [hunk 3]
```diff
@@ -54,234 +42,224 @@ class DataConfig:
...
-        do_import = config.import_directory is not None
-        do_dynamics = config.dynamic_updates_from is not None
-        assert (int(do_generate) + int(do_import) + int(do_dynamics)) == 1
-        if do_generate:
-            self.generate()
-
-        elif do_import:
-            file_utils.import_data_to_storage(self.storage,
-                                              config.import_directory)
+    n_sorted: int
```
Comment (ebianchi) [inline dair_pll/dataset_management.py:169]: It isn't immediately clear to me what `n_sorted` means.  A comment would be helpful here.  After looking at more of the code, it looks like it's a way to keep track of if:
- a dataset has been divvied up into train/valid/test sets, and
- a dynamic dataset has gotten an update and requires processing new trajectories.

The other attribute names are straightforward, but I'll leave it to you to decide whether to comment the other five for consistency.

File: dair_pll/dataset_management.py [hunk 4]
```diff
@@ -54,234 +42,224 @@ class DataConfig:
...
 
+    def __init__(self, storage: str, config: DataConfig,
+                 initial_split: Optional[Tuple[Tensor, Tensor, Tensor]] = None,
+                 use_ground_truth: bool = False) -> None:
+        """s
```
Comment (ebianchi) [inline dair_pll/dataset_management.py]: Extraneous 's' or a placeholder?

File: dair_pll/dataset_management.py [hunk 5]
```diff
@@ -54,234 +42,224 @@ class DataConfig:
...
-        if do_generate:
-            self.generate()
-
-        elif do_import:
-            file_utils.import_data_to_storage(self.storage,
-                                              config.import_directory)
+    n_sorted: int
 
+    def __init__(self, storage: str, config: DataConfig,
+                 initial_split: Optional[Tuple[Tensor, Tensor, Tensor]] = None,
```
Comment (ebianchi) [inline dair_pll/dataset_management.py:173]: It took me a while to realize these weren't actually the same thing, but here's my original interpretation: 
 The `config` already has attributes `train_fraction`, `valid_fraction`, and `test_fraction`, yet there's another argument here `initial_split` that implies the same information.  I see now that `initial_split` is a tuple of the actual indices of the dataset getting allotted to the train/valid/test sets, and not the fractions.  I think the variable names are fine, but interpretability could be aided by a function comment describing the `__init__` arguments.

File: dair_pll/dataset_management.py [hunk 6]
```diff
@@ -54,234 +42,224 @@ class DataConfig:
...
-    t_prediction: int = 1
-    r"""Number of future steps to use during training/evaluation, ``>= 1``\ ."""
-    generation_config: Optional[DataGenerationConfig] = None
-    """Optionally, signals generation of data from given system."""
-    import_directory: Optional[str] = None
-    """Alternatively, signals data import from separate directory."""
-    dynamic_updates_from: Optional[int] = None
-    """Alternatively, loads dynamically, but blocks on initial size set."""
+    slice_config: TrajectorySliceConfig = field(
+        default_factory=TrajectorySliceConfig)
```
Comment (ebianchi) [inline dair_pll/dataset_management.py:46]: Comment description?

File: dair_pll/experiment.py [hunk 7]
```diff
@@ -23,13 +22,34 @@
...
 from dair_pll.system import System, SystemSummary
 from dair_pll.wandb_manager import WeightsAndBiasesManager
 
+
+@dataclass
+class TrainingState:
+    """Dataclass to store a complete summary of the state of training
+    process."""
+    trajectory_set_split: Tuple[Tensor, Tensor, Tensor]
+    """Which trajectories are in train/valid/test sets."""
```
Comment (ebianchi) [inline dair_pll/experiment.py]: Suggest including the word "indices" in "Which trajector[y indices] are in train/valid/test sets."

File: dair_pll/experiment.py [hunk 8]
```diff
@@ -181,8 +143,6 @@ class SupervisedLearningExperiment(ABC):
     """
     config: SupervisedLearningExperimentConfig
     """Configuration of the experiment."""
-    data_manager: SystemDataManager
```
Comment (ebianchi) [inline dair_pll/experiment.py:184]: Why did the `data_manager` attribute get removed from this class?    Now it has to be an argument for the method `per_epoch_evaluation`, which used to just internally retrieve the saved attribute.

I'm not necessarily asking for a change here, I'm mostly curious about what is to me a non-intuitive update.

File: dair_pll/experiment.py [hunk 9]
```diff
@@ -517,84 +542,90 @@ def train(
 
             Best-seen validation set loss.
 
-            Fully-trained system, with parameters corresponding to
+            Fully-trained system, with parameters corresponding to best-seen
+            validation loss.
```
Comment (ebianchi) [inline dair_pll/experiment.py:560]: Need to add comment for the newly-added `ExperimentDataManager` output argument.

File: dair_pll/experiment.py [hunk 10]
```diff
@@ -709,7 +740,8 @@ def to_json(possible_tensor: Union[float, List, Tensor]) -> \
         stats.update(summary_stats)
         return stats
 
-    def evaluation(self, learned_system: System) -> StatisticsDict:
+    def _evaluation(self, learned_system: System,
+                    data_manager: ExperimentDataManager) -> StatisticsDict:
```
Comment (ebianchi) [inline dair_pll/experiment.py]: Function comment needs to list the newly added `data_manager` argument.

File: dair_pll/file_utils.py [hunk 11]
```diff
@@ -205,3 +233,84 @@ def get_hyperparameter_filename(storage_name: str, study_name: str) -> str:
     """Absolute path of optimized hyperparameters for a study"""
     return path.join(study_dir(storage_name, study_name),
                      HYPERPARAMETERS_FILENAME)
+
+
+def load_binary(filename: str, load_callback: Callable[[BinaryIO], Any]) -> Any:
```
Comment (ebianchi) [inline dair_pll/file_utils.py:238]: Is there a reason this has `load_callback` as an argument, as opposed to just internally using `pickle.load` since that's the only thing that gets passed into these function calls?  Unlike the new `load_string()` function, `load_binary()` doesn't do anything differently based on what gets passed into `loss_callback`.

File: dair_pll/file_utils.py [hunk 12]
```diff
@@ -205,3 +233,84 @@ def get_hyperparameter_filename(storage_name: str, study_name: str) -> str:
...
+    with open(filename, 'r', encoding='utf8') as file:
+        if load_callback:
+            value = load_callback(file)
+        else:
+            value = file.read()
+    return value
+
+
+def save_binary(filename: str, value: Any,
+                save_callback: Callable[[Any, BinaryIO], None]) -> None:
```
Comment (ebianchi) [inline dair_pll/file_utils.py:257]: Same thought about whether `save_callback` should get eliminated and the function could just use `pickle.dump` internally.

File: dair_pll/multibody_learnable_system.py [hunk 13]
```diff
@@ -96,8 +96,7 @@ def generate_updated_urdfs(self) -> Dict[str, str]:
         for urdf_name, new_urdf_string in new_urdf_strings.items():
             old_urdf_filename = path.basename(old_urdfs[urdf_name])
             new_urdf_path = path.join(self.output_urdfs_dir, old_urdf_filename)
-            with open(new_urdf_path, 'w', encoding="utf8") as new_urdf_file:
-                new_urdf_file.write(new_urdf_string)
+            file_utils.save_string(new_urdf_path, new_urdf_string)
```
Comment (ebianchi) [inline dair_pll/multibody_learnable_system.py:99]: This (and a similar spot in `urdf_utils.py` that wrote a string) is a great cleanup!

File: examples/contactnets_simple.py [hunk 14]
```diff
@@ -66,10 +70,7 @@
     CUBE_SYSTEM: CUBE_SAMPLER_RANGE,
     ELBOW_SYSTEM: ELBOW_SAMPLER_RANGE
 }
-TRAJ_LENS = {CUBE_SYSTEM: 80, ELBOW_SYSTEM: 120}
-
-# dynamic load configuration.
-DYNAMIC_UPDATES_FROM = 4
+TRAJECTORY_DURATIONS = {CUBE_SYSTEM: 80, ELBOW_SYSTEM: 120}
```
Comment (ebianchi) [inline examples/contactnets_simple.py]: Another location where "duration" is used to describe a number of time steps, as opposed to what is my interpretation of a duration as a length of time (in time units).

File: dair_pll/study.py
Comment (ebianchi) [inline dair_pll/study.py:1]: These changes look fine upon inspection, though I'm documenting here that I haven't tested this file, since it's currently a WIP.

File: dair_pll/wandb_manager.py [hunk 15]
```diff
@@ -52,12 +52,24 @@ def __init__(self, run_name: str, directory: str,
...
+        Todo:
+            Because we manage local storage separate from remote,
+            it's possible that a crash at just the wrong time could cause
+            ``dair_pll`` to think it's resuming a run, while W&B never
+            got notified of the run. To handle this discrepancy, we set
+            ``resume="allow"``\ , which may cause unexpected behavior if a
+            new run is given the same name as an old run on W&B. Some way of
+            rectifying this behavior, such as moving the resume detection
+            fully to W&B, should eventually be implemented.
+        """
```
Comment (ebianchi) [inline dair_pll/wandb_manager.py]: Putting an in-line comment here to flag the warning statement about a W&B issue, related to a worse one I describe in more detail in the overall review comment.

**Review by ebianchi — CHANGES_REQUESTED** [review]
This is going to be a very useful feature to have!  Overall I appreciate the thoughtful placement of each of the moving parts in this implementation, and testing it with `contactnets_simple.py` produced the expected outcomes.  I left some minor thoughts in comments.

One larger concern has to do with the local versus remote W&B storage and what happens if you try to reuse an experiment run name.  Though this may look like it could have been an issue starting with the move from Tensorboard to W&B (before this additional experiment saving and resumption functionality), it's actually a new issue now that `wandb.init` is called with `resume="allow"` as in this PR.

### Deeper description of the danger of reusing an experiment run name
You noted one potential issue that could arise from local versus remote W&B storage in a `Todo` comment in the class definition of `WeightsAndBiasesManager`.  However I've found another issue, and unfortunately a use case which could be fairly common and not just with a "crash at just the wrong time".  Here's a step-by-step way to reproduce the problem:

1. Create and start training an experiment run, with name `test_run`.
2. Stop the training process at epoch `n`.  Observe on the W&B run dashboard the progress of the learned parameters and loss, etc, through epoch `n`.
3. Clear the storage directory, then create and start an experiment run with the same name `test_run`.

At this point, the W&B run dashboard will still show the old results from the first time `test_run` was created through epoch `n`, then the new results starting with epoch `n+1`.  From the beginning, the command line will print `Resuming run test_run` and claim the results can be viewed on its dashboard, though all the new results through epoch `n` are ignored in favor of the old results.

Here's an example of some plots on the W&B run dashboard page when this happens.  The learning process made it through epoch 17 the first time the run was created.  Thus you'll see in all the below plots (except `cube_body_m` which coincidentally looks fairly smooth), there's a jump between epochs 17 and 18, since those don't actually correspond to the same training run.
image

File: dair_pll/dataset_generation.py (hunk 1 above)
Comment (mshalm) [reply dair_pll/dataset_generation.py]: It's to periodically save progress, especially for `n_pop` that would take a long time to compute (e.g. hours).

To aid in understanding, I will move `30` to a module member variable with an interpretable name and comment.

File: dair_pll/dataset_management.py (hunk 5 above)
Comment (mshalm) [reply dair_pll/dataset_management.py:173]: On my local storage I'm seeing that `__init__` already has comments, but from this discussion I don't think the `initial_split` comment is particularly clear, so I'll update it.

File: dair_pll/experiment.py (hunk 8 above)
Comment (mshalm) [reply dair_pll/experiment.py:184]: Yeah, in retrospect, an unhelpful/unnecessary change. Will revert.

File: dair_pll/file_utils.py (hunk 11 above)
Comment (mshalm) [reply dair_pll/file_utils.py:238]: It's true that pickle files are the only binary files we load manually at the moment. I'm content to leave it as-is to have some structure for when/if there may be other types of binary files we want to load. But if you feel strongly I'm fine refactoring to a reasonable instance of `load_pickle`

File: dair_pll/dataset_management.py (hunk 3 above)
Comment (mshalm) [reply dair_pll/dataset_management.py:169]: Added some documentation here

**Comment by mshalm** [comment]
All comments addressed in the latest commits.

The biggest design change was made to address W&B's inability to start runs with the same name as a previously deleted run.

To compensate for this, the `wandb` training run ID is related to the `dair_pll` experiment `run_name` by
```
wandb_run_id = f"{run_name}_{time.time_ns()}"
```
where the timestamp comes from the time the run was started. This label is added to the `TrainingState` to enable restarting partially completed W&B runs.

File: dair_pll/file_utils.py (hunk 11 above)
Comment (ebianchi) [reply dair_pll/file_utils.py:238]: I don't feel strongly -- your reasoning makes sense, so let's leave it as-is.

**Review by ebianchi — APPROVED** [review]
Everything looks good and works as expected.  Great improvement to add the timestamp to the end of the `wandb` IDs to elegantly restart old versus create new experiment runs.  All approved on my end.
